pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java

```diff
@@ -451,7 +554,44 @@ private boolean isUserProvidedProducerName(Producer producer){
         return producer.isUserProvidedProducerName() && !producer.getProducerName().startsWith(replicatorPrefix);
     }
 
-    protected abstract void handleProducerRemoved(Producer producer);
+
+    @Override
+    public void removeProducer(Producer producer) {
+        checkArgument(producer.getTopic() == this);
+
+        if (producers.remove(producer.getProducerName(), producer)) {
+            handleProducerRemoved(producer);
```

Comment:
shouldn't we need lock here else it may create a race condition and  producer with `WaitForExclusive ` may wait forever.

Reply:
No lock needed there so far. Other changes are needed for `WaitForExclusive`, it's not implemented in this PR